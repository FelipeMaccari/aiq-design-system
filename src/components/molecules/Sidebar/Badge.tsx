import React from 'react'
import PropTypes from 'prop-types'

import { Text } from '../../atoms/Text'
import { Flex } from '../../atoms/Flex'

export const Badge = ({ children }) => {
  return (
    <Flex
      backgroundColor='#ff4d4f'
      variant='centralized'
      padding='1px 8px'
      marginRight='16px'
      borderRadius='12px'
    >
      <Text fontSize='small' color='white'>
        {children}
      </Text>
    </Flex>
  )
}

Badge.propTypes = {
  children: PropTypes.any
}
