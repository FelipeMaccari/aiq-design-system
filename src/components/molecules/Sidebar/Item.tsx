import React, { useState, useMemo } from 'react'
import PropTypes from 'prop-types'
import { MdExpandMore, MdExpandLess } from 'react-icons/md'
import styled from 'styled-components'

import { Text } from '../../atoms/Text'
import { Flex } from '../../atoms/Flex'
import { Link } from '../../atoms/Link'
import { Icon } from '../../atoms/Icon'

const ItemStyled = styled.li`
  &:hover {
    cursor: pointer;
  }
  a {
    padding: 16px 24px;
    display: flex;
    flex-direction: row;
  }
`
interface SubItens {
  isOpen: boolean
}

const SubItens = styled.ul<SubItens>`
  margin: 0;
  padding: 0;
  background: #f6f6f6;

  &.hide {
    overflow: hidden;
    max-height: 0;
    padding-top: 0;
    padding-bottom: 0;
    margin-top: 0;
    margin-bottom: 0;
    transition: all 0.2s ease;
    will-change: transform;
  }

  &.show {
    max-height: 360px;
    overflow: hidden;
    transition: all 0.5s ease;
    will-change: transform;
  }
  a {
    color: #434343;
    padding: 12px 24px;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }
`

const Badge = ({ children }) => {
  return (
    <Flex
      backgroundColor='#ff4d4f'
      variant='centralized'
      padding='1px 8px'
      marginRight='16px'
      borderRadius='12px'
    >
      <Text fontSize='small' color='white'>
        {children}
      </Text>
    </Flex>
  )
}

Badge.propTypes = {
  children: PropTypes.any
}

interface ItemProps {
  item: any
}

export const Item: React.FC<ItemProps> = ({ item, ...props }) => {
  const [isOpen, setIsOpen] = useState(false)

  function computeBadgeAllItens(iten) {
    let value = 0
    if (iten.itens) {
      iten.itens.forEach(subItem => {
        if (subItem.badge) value += subItem.badge
      })
    }
    return value
  }

  const badgeAllItens = useMemo(() => computeBadgeAllItens(item), [item])

  function changeVisibilitySubItem() {
    setIsOpen(!isOpen)
  }

  if (item.itens && item.itens.length > 0) {
    return (
      <ItemStyled {...props}>
        <Flex flexDirection='column'>
          <Flex
            flexDirection='row'
            alignItems='center'
            justifyContent='space-between'
            padding='16px 24px'
            onClick={changeVisibilitySubItem}
          >
            <Flex>
              <Icon marginRight='18px' color='primary'>
                {item.icon}
              </Icon>
              {isOpen ? (
                <Text cursor='pointer' color='primary' fontWeight='medium'>
                  {item.name}
                </Text>
              ) : (
                <Text cursor='pointer' color='darkerGrey'>
                  {item.name}
                </Text>
              )}
            </Flex>
            <Flex>
              {!isOpen && badgeAllItens ? (
                <Badge>{badgeAllItens}</Badge>
              ) : (
                <Flex />
              )}

              <Icon size='18px' color='grey'>
                {isOpen ? <MdExpandLess /> : <MdExpandMore />}
              </Icon>
            </Flex>
          </Flex>
          <SubItens className={`${isOpen ? 'show' : 'hide'}`} isOpen={isOpen}>
            {item.itens.map((subItem, index) => (
              <li key={index}>
                <Link
                  variant={subItem.type ? subItem.type : 'internal'}
                  href={subItem.href}
                >
                  <Text cursor='pointer' fontSize='medium'>
                    {subItem.name}
                  </Text>
                  {subItem.badge && <Badge>{subItem.badge}</Badge>}
                </Link>
              </li>
            ))}
          </SubItens>
        </Flex>
      </ItemStyled>
    )
  }

  return (
    <ItemStyled {...props}>
      <Link variant={item.type ? item.type : 'internal'} href={item.href}>
        <Icon marginRight='18px' color='primary'>
          {item.icon}
        </Icon>
        <Text cursor='pointer' color='darkerGrey'>
          {item.name}
        </Text>
      </Link>
    </ItemStyled>
  )
}

Item.propTypes = {
  item: PropTypes.any.isRequired
}
