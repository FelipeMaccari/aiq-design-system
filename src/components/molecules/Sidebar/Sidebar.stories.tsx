import React from 'react'
import { MdHome, MdBookmark, MdStorage } from 'react-icons/md'
import { BrowserRouter as Router } from 'react-router-dom'
import { Sidebar } from './Sidebar'
import { withKnobs, boolean } from '@storybook/addon-knobs'

export default {
  component: Sidebar,
  title: 'molecules/Sidebar',
  decorators: [withKnobs]
}

export const Basic: React.FC = () => {
  const sidebarData = {
    user: {
      name: 'juliano',
      email: 'teste@aiqfome.com'
    },
    itens: [
      {
        icon: <MdHome />,
        name: 'dashboard',
        href: '/dashboard'
      },
      {
        icon: <MdBookmark />,
        name: 'pedidos',
        href: '/pedidos'
      },
      {
        icon: <MdStorage />,
        name: 'lojas',
        itens: [
          {
            name: 'gerenciar lojas',
            href: '/gerenciar_lojas'
          },
          {
            name: 'adicionar loja',
            href: '/adicionar_lojas'
          },
          {
            name: 'lojas prospectas',
            href: '/gerenciar_lojas',
            badge: 18
          }
        ]
      },
      {
        icon: <MdStorage />,
        name: 'lojas',
        itens: [
          {
            name: 'gerenciar lojas',
            href: '/gerenciar_lojas'
          },
          {
            name: 'adicionar loja',
            href: '/adicionar_lojas'
          },
          {
            name: 'lojas prospectas',
            href: '/gerenciar_lojas'
          }
        ]
      },
      {
        icon: <MdBookmark />,
        name: 'pedidos',
        href: '/pedidos'
      },
      {
        icon: <MdBookmark />,
        name: 'pedidos',
        href: '/pedidos'
      },
      {
        icon: <MdBookmark />,
        name: 'pedidos',
        href: 'https://aiqfome.com/',
        type: 'external'
      }
    ]
  }

  return (
    <Router>
      <Sidebar opened={boolean('opened', true)} data={sidebarData} />
    </Router>
  )
}
