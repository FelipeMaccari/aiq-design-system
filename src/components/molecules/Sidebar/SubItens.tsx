import React from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import { MdKeyboardArrowRight } from 'react-icons/md'
import { Text } from '../../atoms/Text'
import { Flex } from '../../atoms/Flex'
import { Link } from '../../atoms/Link'
import { Icon } from '../../atoms/Icon'

import { Badge } from './Badge'

interface Props {
  item?: any
  itemOpened?: boolean
  sidebarOpened?: boolean
}

const SubItensStyled = styled(Flex)<Props>`
  ul {
    margin: 0;
    padding: 0;
    list-style: none;
    padding-bottom: 24px;
  }

  ${({ sidebarOpened }) => {
    if (sidebarOpened) {
      return css`
        background: #f6f6f6;

        &.hide {
          overflow: hidden;
          max-height: 0;
          padding-top: 0;
          padding-bottom: 0;
          margin-top: 0;
          margin-bottom: 0;
          transition: all 0.2s ease;
          will-change: transform;
        }

        &.show {
          max-height: 360px;
          overflow: hidden;
          transition: all 0.5s ease;
          will-change: transform;
        }

        a {
          color: #434343;
          padding: 12px 24px;
          display: flex;
          flex-direction: row;
          justify-content: space-between;

          &:hover {
            background: ${({ theme }) => theme.colors.grey};
          }
        }
      `
    }
    return css<Props>`
      h3 {
        padding-top: 16px;
        color: #434343;
        padding: 8px 16px;
        font-size: 16px;
      }

      a {
        font-size: 14px;
        color: #434343;
        font
        padding: 6px 0px;
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        &:hover {
          background: ${({ theme }) => theme.colors.grey};
        }
      }

      ${({ itemOpened }) => {
        if (itemOpened) {
          return css`
            position: absolute;
            width: 240px;
            top: 0;
            right: -240px;
            background: #ffff;
            border-radius: 4px;
            box-shadow: 0px 0px 15px #0000001a;
          `
        }
        return css`
          max-height: 0;
          overflow: hidden;
        `
      }}
    `
  }}
`

export const SubItens = ({ item, sidebarOpened, itemOpened }) => {
  return (
    <SubItensStyled
      flexDirection='column'
      className={`${itemOpened ? 'show' : 'hide'}`}
      sidebarOpened={sidebarOpened}
      itemOpened={itemOpened}
    >
      {!sidebarOpened && <h3>{item.name}</h3>}
      {item.itens && (
        <ul>
          {item.itens.map((subItem, index) => (
            <li key={index}>
              <Link
                variant={subItem.type ? subItem.type : 'internal'}
                href={subItem.href}
              >
                <Flex>
                  <Icon paddingLeft='16px' marginRight='8px'>
                    <MdKeyboardArrowRight />
                  </Icon>
                  <Text cursor='pointer' fontSize='medium'>
                    {subItem.name}
                  </Text>
                </Flex>

                {subItem.badge && <Badge>{subItem.badge}</Badge>}
              </Link>
            </li>
          ))}
        </ul>
      )}
    </SubItensStyled>
  )
}

SubItens.propTypes = {
  item: PropTypes.any,
  itemOpened: PropTypes.bool,
  sidebarOpened: PropTypes.bool
}
