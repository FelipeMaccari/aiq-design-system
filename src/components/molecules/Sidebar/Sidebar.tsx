import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { Drawer } from '../../atoms/Drawer'
import { Avatar } from '../../atoms/Avatar'
import { Text } from '../../atoms/Text'
import { Divider } from '../../atoms/Divider'
import { Flex } from '../../atoms/Flex'
import { Item } from './Item'

type SidebarData = {
  user: any
  itens: any
}

export interface Props {
  data: SidebarData
  opened?: boolean
}

const Itens = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
`

export const Sidebar: React.FC<Props> = ({
  data,
  opened = false,
  ...props
}) => {
  return (
    <Drawer
      width='100%'
      maxWidth='340px'
      boxShadow='0px 3px 15px #0000001a'
      variation='left'
      opened={opened}
      {...props}
    >
      {data && (
        <>
          <Flex padding='24px' alignItems='center'>
            <Avatar alt={data.user.name} mx='12px' />
            <Flex flexDirection='column'>
              <Text color='almostBlack' fontWeight='semiBold'>
                {data.user.name}
              </Text>
              <Text color='darkGrey' fontSize='xsmall'>
                {data.user.email}
              </Text>
            </Flex>
          </Flex>
          <Divider width='100%' marginBottom='16px' />
          <Itens>
            {data.itens.map((item, index) => (
              <Item key={index} item={item} />
            ))}
          </Itens>
        </>
      )}
    </Drawer>
  )
}

Sidebar.propTypes = {
  data: PropTypes.any.isRequired,
  opened: PropTypes.bool
}
