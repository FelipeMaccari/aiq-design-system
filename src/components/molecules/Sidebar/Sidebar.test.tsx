import React from 'react'
import renderer from 'react-test-renderer'

import { Sidebar } from './Sidebar'

describe('must match with the previous snapshot', () => {
  test('snapshot renders', () => {
    const component = renderer.create(<Sidebar />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
