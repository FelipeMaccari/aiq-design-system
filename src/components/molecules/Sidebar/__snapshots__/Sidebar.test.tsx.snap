// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must match with the previous snapshot snapshot renders 1`] = `
<div
  className="sc-AxjAm jYazFB sc-fzpans eWDhsX"
>
  <div
    className="sc-AxhUy dDmHvy"
  >
    <div
      className="sc-AxhCb jkHhfy"
    >
      <span
        className="sc-AxirZ hCmXwK"
        color="primary"
        fontWeight="bold"
      >
        
      </span>
    </div>
    <div
      className="sc-AxhUy jzCCTg"
    >
      <span
        className="sc-AxirZ nOQeL"
        color="almostBlack"
        fontWeight="semiBold"
      >
        juliano
      </span>
      <span
        className="sc-AxirZ kDNAxe"
        color="darkGrey"
        fontSize="xsmall"
      >
        test@aiqfome.com
      </span>
    </div>
  </div>
  <hr
    className="sc-AxgMl tfALA"
    height="1px"
    width="100%"
  />
  <ul
    className="sc-fzplWN kyjGRr"
  >
    <li
      className="sc-fzozJi gcECS"
    >
      <a
        className="sc-AxheI juDcHF"
        href="/dashboard"
        onClick={[Function]}
      >
        <div
          className="sc-AxhUy sc-AxmLO kmVXqP"
          color="primary"
        >
          <svg
            fill="currentColor"
            height={18}
            size={18}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width={18}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </div>
        <span
          className="sc-AxirZ bAjyB"
          color="darkerGrey"
          cursor="pointer"
        >
          dashboard
        </span>
      </a>
    </li>
    <li
      className="sc-fzozJi gcECS"
    >
      <a
        className="sc-AxheI juDcHF"
        href="/pedidos"
        onClick={[Function]}
      >
        <div
          className="sc-AxhUy sc-AxmLO kmVXqP"
          color="primary"
        >
          <svg
            fill="currentColor"
            height={18}
            size={18}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width={18}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </div>
        <span
          className="sc-AxirZ bAjyB"
          color="darkerGrey"
          cursor="pointer"
        >
          bookmark
        </span>
      </a>
    </li>
    <li
      className="sc-fzozJi gcECS"
    >
      <div
        className="sc-AxhUy jzCCTg"
      >
        <div
          className="sc-AxhUy ewRgww"
          onClick={[Function]}
        >
          <div
            className="sc-AxhUy hWpvLO"
          >
            <div
              className="sc-AxhUy sc-AxmLO kmVXqP"
              color="primary"
            >
              <svg
                fill="currentColor"
                height={18}
                size={18}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width={18}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 20h20v-4H2v4zm2-3h2v2H4v-2zM2 4v4h20V4H2zm4 3H4V5h2v2zm-4 7h20v-4H2v4zm2-3h2v2H4v-2z"
                />
              </svg>
            </div>
            <span
              className="sc-AxirZ bAjyB"
              color="darkerGrey"
              cursor="pointer"
            >
              store
            </span>
          </div>
          <div
            className="sc-AxhUy hWpvLO"
          >
            <div
              className="sc-AxhUy gvsbX"
            >
              <span
                className="sc-AxirZ ezsgln"
                color="white"
                fontSize="small"
              >
                18
              </span>
            </div>
            <div
              className="sc-AxhUy sc-AxmLO rnzlv"
              color="grey"
            >
              <svg
                fill="currentColor"
                height={18}
                size={18}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width={18}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </div>
          </div>
        </div>
        <ul
          className="sc-fzoLsD cRaTUv hide"
        >
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
            </a>
          </li>
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
            </a>
          </li>
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
              <div
                className="sc-AxhUy gvsbX"
              >
                <span
                  className="sc-AxirZ ezsgln"
                  color="white"
                  fontSize="small"
                >
                  18
                </span>
              </div>
            </a>
          </li>
        </ul>
      </div>
    </li>
    <li
      className="sc-fzozJi gcECS"
    >
      <div
        className="sc-AxhUy jzCCTg"
      >
        <div
          className="sc-AxhUy ewRgww"
          onClick={[Function]}
        >
          <div
            className="sc-AxhUy hWpvLO"
          >
            <div
              className="sc-AxhUy sc-AxmLO kmVXqP"
              color="primary"
            >
              <svg
                fill="currentColor"
                height={18}
                size={18}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width={18}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 20h20v-4H2v4zm2-3h2v2H4v-2zM2 4v4h20V4H2zm4 3H4V5h2v2zm-4 7h20v-4H2v4zm2-3h2v2H4v-2z"
                />
              </svg>
            </div>
            <span
              className="sc-AxirZ bAjyB"
              color="darkerGrey"
              cursor="pointer"
            >
              itens
            </span>
          </div>
          <div
            className="sc-AxhUy hWpvLO"
          >
            <div
              className="sc-AxhUy hWpvLO"
            />
            <div
              className="sc-AxhUy sc-AxmLO rnzlv"
              color="grey"
            >
              <svg
                fill="currentColor"
                height={18}
                size={18}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width={18}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </div>
          </div>
        </div>
        <ul
          className="sc-fzoLsD cRaTUv hide"
        >
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
            </a>
          </li>
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
            </a>
          </li>
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
            </a>
          </li>
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
            </a>
          </li>
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
            </a>
          </li>
          <li>
            <a
              className="sc-AxheI juDcHF"
              href="/subItem"
              onClick={[Function]}
            >
              <span
                className="sc-AxirZ cnzlxO"
                cursor="pointer"
                fontSize="medium"
              >
                Sub Item
              </span>
            </a>
          </li>
        </ul>
      </div>
    </li>
    <li
      className="sc-fzozJi gcECS"
    >
      <a
        className="sc-Axmtr fGBbby"
        href="https://aiqfome.com/"
      >
        <div
          className="sc-AxhUy sc-AxmLO kmVXqP"
          color="primary"
        >
          <svg
            fill="currentColor"
            height={18}
            size={18}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width={18}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </div>
        <span
          className="sc-AxirZ bAjyB"
          color="darkerGrey"
          cursor="pointer"
        >
          aiqfome
        </span>
      </a>
    </li>
  </ul>
</div>
`;
